### PR TITLE
Cleanup metadata subcommand, remove unused parameters

### PR DESCRIPTION
**Pull request by vicancy**

### CODE REVIEW

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs [hunk 1]
```diff
@@ -946,6 +872,22 @@ private static bool TryParseYamlMetadataFile(string metadataFileName, out Metada
...
         }
 
+        private async Task<Project> GetProjectJsonProjectAsync(string path)
+        {
+            try
+            {
+                Logger.LogVerbose($"Loading project {path}", file: path);
+                var workspace = new ProjectJsonWorkspace(path);
+                var project = workspace.CurrentSolution.Projects.FirstOrDefault(p => p.FilePath == Path.GetFullPath(path));
+                return await Task.FromResult(project);
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs]: hehe

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataUtility.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        }
+
+        public static string Normalize(string path)
+        {
+            if (string.IsNullOrEmpty(path))
+            {
+                return path;
+            }
+
+            return Path.GetFullPath(Path.Combine(EnvironmentContext.BaseDirectory, path)).ToNormalizedPath();
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataUtility.cs]: >Path.GetFullPath [](start = 19, length = 16)

can use ToNormalizedFullPath

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs [hunk 3]
```diff
@@ -164,169 +164,82 @@ internal static MetadataItem GenerateYamlMetadata(Compilation compilation, IAsse
...
-                }
-            }, 60);
-
-            // Load additional projects out if it is not contained in expanded solution
-            projects = projects.Except(projectCache.Keys).Distinct();
+                }, 60);
+            }
 
-            await projects.ForEachInParallelAsync(async path =>
+            if (_files.TryGetValue(FileType.ProjectJsonProject, out var pjprj))
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs]: `pjprj`...

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs [hunk 4]
```diff
@@ -20,50 +19,52 @@ namespace Microsoft.DocAsCode.Metadata.ManagedReference
...
+                throw new ArgumentNullException(nameof(input.OutputFolder), "Output folder must be specified");
+            }
+
+            if (input.Files == null || input.Files.Count == 0)
+            {
+                Logger.Log(LogLevel.Warning, "No source project or file to process, exiting...");
+                return;
+            }
+
+            _files = ExtractMetadataUtility.GetInputs(input.Files).GroupBy(f => f.Type).ToDictionary(s => s.Key, s => s.Distinct().ToList());
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs]: >Distinct [](start = 120, length = 8)

distinct fileInfo needs to overwrite equals and hashcode?

Comment (vicancy) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs]: yes, absolutely:thumbsup: 

---
In reply to: [118815303](https://github.com/dotnet/docfx/pull/1707#discussion_r118815303) [](ancestors = 118815303)

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs [hunk 5]
```diff
@@ -76,27 +79,26 @@ public ExtractMetadataWorker(ExtractMetadataInputModel input, bool rebuild, bool
...
-
             try
             {
-                foreach (var pair in validInput.Items)
+                if (_files.TryGetValue(FileType.NotSupported, out List<FileInfo> unsupportedFiles))
                 {
-                    var inputs = pair.Value;
-                    var outputFolder = pair.Key;
-                    await SaveAllMembersFromCacheAsync(inputs, outputFolder, _rebuild);
+                    Logger.LogWarning(GetPrintableFileList(unsupportedFiles));
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs]: >LogWarning [](start = 27, length = 10)

besides file lise, add some warning message?

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs [hunk 6]
```diff
@@ -164,169 +166,81 @@ internal static MetadataItem GenerateYamlMetadata(Compilation compilation, IAsse
...
                             {
-                                var path = StringExtension.ToNormalizedFullPath(inputFilePath);
-                                validFilePath.Add(path);
+                                projectCache.GetOrAdd(projectFile.NormalizedPath, s => project);
                             }
                             else
                             {
-                                var value = string.Join(",", SupportedExtensions);
-                                Logger.Log(LogLevel.Warning, $"File {inputFilePath} is not supported, supported file extension are: {value}. The file will be ignored.");
+                                Logger.Log(LogLevel.Warning, $"Project {projectFile.RawPath} inside solution {path} is ignored, supported projects are csproj and vbproj.");
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs]: >.Log(LogLevel.Warning [](start = 38, length = 21)

direct log warning?

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs [hunk 7]
```diff
@@ -164,169 +166,81 @@ internal static MetadataItem GenerateYamlMetadata(Compilation compilation, IAsse
...
-                        {
-                            projectCache.GetOrAdd(StringExtension.ToNormalizedFullPath(project.FilePath), s => project);
-                        }
-                        else
-                        {
-                            var value = string.Join(",", SupportedExtensions);
-                            Logger.Log(LogLevel.Warning, $"Project {filePath} inside solution {path} is not supported, supported file extension are: {value}. The project will be ignored.");
-                        }
+                        var project = await GetProjectAsync(path);
+                        projectCache.GetOrAdd(path, project);
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs]: >GetOrAdd [](start = 37, length = 8)

use getOrAdd without checking key? when cannot get, add will call GetProjectAsync function?

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/FileInfo.cs [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+namespace Microsoft.DocAsCode.Metadata.ManagedReference
+{
+    using System;
+    using System.IO;
+
+    using Microsoft.DocAsCode.Common;
+    using Microsoft.DocAsCode.Plugins;
+
+    internal class FileInfo
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/FileInfo.cs]: >FileInfo [](start = 19, length = 8)

same name with System.IO.FileInfo

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs [hunk 9]
```diff
@@ -346,13 +260,35 @@ private static ExtractMetadataInputModel ValidateInput(ExtractMetadataInputModel
...
+            if (_files.TryGetValue(FileType.CSSourceCode, out var cs))
+            {
+                csFiles.AddRange(cs.Select(s => s.NormalizedPath));
+                documentCache.AddDocuments(csFiles);
+            }
+
+            if (_files.TryGetValue(FileType.VBSourceCode, out var vb))
+            {
+                vbFiles.AddRange(vb.Select(s => s.NormalizedPath));
+
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs]: minor: remove empty line

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs [hunk 10]
```diff
@@ -346,13 +260,35 @@ private static ExtractMetadataInputModel ValidateInput(ExtractMetadataInputModel
...
+                documentCache.AddDocuments(vbFiles);
+            }
+
+            if (_files.TryGetValue(FileType.Assembly, out var asm))
+            {
+                assemblyFiles.AddRange(asm.Select(s => s.NormalizedPath));
+
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs]: minor: remove empty line

**Comment by hellosnow** [comment]
>                                 Logger.Log(LogLevel.Warning, $"Duplicate member {s.Name} is found from {existingMetadata.Source.Path} and {s.Source.Path}, use the one in {existingMetadata.Source.Path} and ignore the one from {s.Source.Path}");

minor: still several Logger.Log(LogLevel.Warning... in this file, also update?

---
Refers to: src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs:764 in 65e490e. [](commit_id = 65e490edcd58140e67e7042316101950ec00d34e, deletion_comment = False)

**Comment by vicancy** [comment]
>                                 Logger.Log(LogLevel.Warning, $"Duplicate member {s.Name} is found from {existingMetadata.Source.Path} and {s.Source.Path}, use the one in {existingMetadata.Source.Path} and ignore the one from {s.Source.Path}");

further refactor is coming, we can leave this to the next code review :)

---
In reply to: [305079133](https://github.com/dotnet/docfx/pull/1707#issuecomment-305079133) [](ancestors = 305079133)

---
Refers to: src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs:764 in 65e490e. [](commit_id = 65e490edcd58140e67e7042316101950ec00d34e, deletion_comment = False)

**Comment by hellosnow** [comment]
:shipit:

**Comment by superyyrrzz** [comment]
:shipit:

**Pull request merged**